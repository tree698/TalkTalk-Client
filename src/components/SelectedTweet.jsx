import React, { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useApiContext } from '../context/ApiContext';
import NewTweetForm from './NewTweetForm';
import TweetCard from './TweetCard';
import Banner from './ui/Banner';
import Button from './ui/Button';
import UsersInConversation from './UsersInConversation';

export default function SelectedTweet() {
  const [tweets, setTweets] = useState([]);
  const [selectedUsername, setSelectedUsersname] = useState(undefined);
  const [error, setError] = useState('');

  const { user, tweetService } = useApiContext();
  const {
    state: { id },
  } = useLocation();

  useEffect(() => {
    id &&
      tweetService
        .getTweets(selectedUsername, id)
        .then((tweets) => setTweets([...tweets]))
        .catch((error) => {
          setError((prev) => error.toString());
          setTimeout(() => {
            setError('');
          }, 3000);
        });

    const stopSync = tweetService.onSync((tweet) => onCreated(tweet));
    return () => stopSync();
  }, [tweetService, user, selectedUsername]);

  const onCreated = (tweet) => {
    setTweets((tweets) => [tweet, ...tweets]);
  };

  const handleDelete = (tweetId) => {
    tweetService
      .deleteTweet(tweetId)
      .then(() =>
        setTweets((tweets) => tweets.filter((tweet) => tweet.id !== tweetId))
      )
      .catch((error) => setError(error.toString()));
  };

  const handleAvatarClick = (username) => {
    setSelectedUsersname((prev) => username);
  };

  return (
    <div className="basis-2/5">
      <Banner text={error} />
      <UsersInConversation tweets={tweets} onAvatarClick={handleAvatarClick} />
      <Button
        text="All Talks"
        onClick={() => setSelectedUsersname((prev) => undefined)}
      />
      {tweets.length === 0 && <h2>No Messages Yet ...</h2>}
      <ul id="tweets">
        {tweets.map((tweet) => (
          <TweetCard
            key={tweet.id}
            tweet={tweet}
            owner={tweet.username === user.username}
            onAvatarClick={handleAvatarClick}
            onDelete={handleDelete}
          />
        ))}
      </ul>

      <div>
        <NewTweetForm />
      </div>
    </div>
  );
}
