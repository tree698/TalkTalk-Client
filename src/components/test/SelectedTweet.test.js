import { Route } from 'react-router-dom';
import { withApiContext, withRouter } from '../../test/utils';
import SelectedTweet from '../SelectedTweet';
import renderer from 'react-test-renderer';
import { fakeTweets as tweets } from '../../test/tweetsData';
import { render, waitFor } from '@testing-library/react';

jest.mock('../UsersInConversation');
jest.mock('../NewTweetForm');
jest.mock('../TweetCard');

describe('SelectedTweet', () => {
  const fakeContext = {
    user: 'fakeUser',
    tweetService: {
      getTweets: jest.fn(),
      onSync: jest.fn(),
      deleteTweet: jest.fn(),
    },
  };

  it('renders correctly', async () => {
    fakeContext.tweetService.getTweets.mockResolvedValue(tweets);

    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<SelectedTweet />} />, {
          state: { id: 'fakeId' },
        }),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders with tweets', async () => {
    fakeContext.tweetService.onSync.mockReturnValue(jest.fn());
    fakeContext.tweetService.getTweets.mockImplementation(() =>
      Promise.resolve(tweets)
    );

    render(
      withApiContext(
        withRouter(<Route path="/" element={<SelectedTweet />} />, {
          state: { id: 'fakeId' },
        }),
        fakeContext
      )
    );

    await waitFor(() => {
      expect(fakeContext.tweetService.onSync).toHaveBeenCalled();
      expect(fakeContext.tweetService.getTweets).toHaveBeenCalled();
    });
  });
});
