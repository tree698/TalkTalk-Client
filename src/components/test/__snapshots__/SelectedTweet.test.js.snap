// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedTweet renders with tweets 1`] = `
<div
  className="basis-1/2 flex flex-col px-10 md:px-12 lg:pl-8 lg:pr-0"
>
  <button
    className="w-full my-2 py-1 text-xs md:text-sm lg:text-base border border-superLightGray rounded-md hover:bg-slate500 hover:text-white hover:border-0 transition-all delay-150 duration-300 ease-in-out"
    data-testid="allTalks"
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <svg
        className="ml-3 md:ml-5 lg:ml-7 mr-2 md:mr-3 lg:mr-5"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
          fillRule="evenodd"
        />
      </svg>
      <p>
        All Talks
      </p>
    </div>
  </button>
  <div
    className="flex-1 overflow-y-auto bg-talkBG bg-cover bg-no-repeat border border-superLightGray rounded-md overflow-hidden"
    data-testid="chatting-box"
    style={
      Object {
        "boxShadow": "7px 7px 5px 0px rgba(124,121,121,1)",
      }
    }
  >
    <ul
      className="h-96 p-1 md:p-2 lg:p-3"
    >
      <div
        className="flex justify-center items-end gap-2 text-base md:text-xl lg:text-2xl font-bold mt-20 md:mt-28 lg:mt-36"
      >
        <p>
          No talk yet...
        </p>
        <svg
          className="text-2xl md:text-4xl lg:text-5xl pb-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M254.156 56.938c-64.144 0-122.393 19.17-165.03 50.718-42.64 31.547-69.845 76.01-69.845 125.5 0 49.492 27.207 93.952 69.845 125.5 6.388 4.727 13.13 9.164 20.188 13.313l-12.657 95.905 90.594-65.75c21.226 4.708 43.67 7.25 66.906 7.25 64.144 0 122.362-19.17 165-50.72C461.794 327.11 489 282.65 489 233.156c0-49.49-27.206-93.95-69.844-125.5-42.638-31.546-100.856-50.718-165-50.718zM157 274.438c12.69 0 23 10.276 23 22.968 0 12.69-10.31 23-23 23s-22.97-10.31-22.97-23 10.28-22.97 22.97-22.97zm96.844 0c12.69 0 22.97 10.276 22.97 22.968-.002 12.69-10.28 23-22.97 23s-23-10.31-23-23 10.31-22.97 23-22.97zm99.844 0c12.69 0 22.968 10.276 22.968 22.968 0 12.69-10.278 23-22.97 23-12.69 0-22.998-10.31-22.998-23s10.31-22.97 23-22.97z"
          />
        </svg>
      </div>
    </ul>
  </div>
</div>
`;
